Microsoft.Toolkit/Extensions/ISettingsStorageHelperExtensions.cs

```diff
@@ -0,0 +1,88 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.Toolkit.Helpers;
+
+namespace Microsoft.Toolkit.Extensions
+{
+    /// <summary>
+    /// Helpers methods for working with <see cref="ISettingsStorageHelper{TKey}"/> implementations.
+    /// </summary>
+    public static class ISettingsStorageHelperExtensions
+    {
+        /// <summary>
+        /// Attempts to read the provided key and return the value.
+        /// If the key is not found, the fallback value will be used instead.
+        /// </summary>
+        /// <typeparam name="TKey">The type of key used to lookup the object.</typeparam>
+        /// <typeparam name="TValue">The type of object value expected.</typeparam>
+        /// <param name="storageHelper">The storage helper instance fo read from.</param>
+        /// <param name="key">The key of the target object.</param>
+        /// <param name="fallback">An alternative value returned if the read fails.</param>
+        /// <returns>The value of the target object, or the fallback value.</returns>
+        public static TValue? GetValueOrDefault<TKey, TValue>(this ISettingsStorageHelper<TKey> storageHelper, TKey key, TValue? fallback = default)
+            where TKey : notnull
+        {
+            try
+            {
+                return storageHelper.Read<TValue>(key);
+            }
+            catch (KeyNotFoundException)
+            {
+                return fallback;
+            }
```

Comment:
I hear you say they are inefficient... but why/how? Like, where do you get that impression from? The goal is syntactic sugar to make it "easier" to use the storage helper by doing the bare minimum exception handling to assist expected usage patterns.

Currently, the interface defines the signature for `Read`, but no `TryRead`. `Read` should throw a `KeyNotFoundException` if the key isn't found. Therefore, the extension helper is the safe try-like method. Are you saying that I should just migrate those features into the `ISettingsStorageHelper` interface, and that would mitigate the efficiency concern?

Reply:
> "I hear you say they are inefficient... but why/how? Like, where do you get that impression from?"

The reason is that throwing (and especially catching) exception involves a **ton** of extra work, and as such it should absolutely be avoided for flow control - they should only ever be used to actually handle unexpected circumstances. In a nutshell, throwing and catching an exception means that you need to:
- Instantiate the exception object
- Gather information on the stack trace (this is particularly expensive)
- Trigger an interrupt
- The runtime then needs to invoke the interrupt handler to handle it
- It then needs to find the catch block that can handle that exception higher in the call stack, if any
- Control needs to then be dispatched back to that control to resume execution

For instance, I've put together a very simple benchmark to demonstrate this, with just a method either using `TryGet` and then returning a default value, or using an indexer and catching the `KeyNotFoundException` in case of failure, using a local `Dictionary<TKey, TValue>` instance. Here's the results:

|   Method |         Mean |      Error |     StdDev |    Ratio |  Gen 0 | Allocated |
|--------- |-------------:|-----------:|-----------:|---------:|-------:|----------:|
|   TryGet |     5.592 ns |  0.0225 ns |  0.0210 ns |     1.00 |      - |         - |
| **TryCatch** | **7,350.567 ns** | 35.7207 ns | 29.8284 ns | **1,314.98** | 0.1450 |     **688 B** |

That is, catching the exception is about **1300x** slower than just handing the logic properly.
It's not just a bit slower, it's incredibly slower unfortunately 😄

> "The goal is syntactic sugar to make it "easier" to use the storage helper by doing the bare minimum exception handling to assist expected usage patterns."

Absolutely, and that's great! I'm just saying that we can do this more efficiently like the BCL does as well 🙂

> "Are you saying that I should just migrate those features into the `ISettingsStorageHelper` interface, and that would mitigate the efficiency concern?"

The way I see it, the easy solution to also keep the interface as compact as possible would be to just swap the two: the interface would only expose the `TryRead` method, which could then be implemented without relying on exception handling, and then the `Read` method would instead be an extension method that would simply build on top of the other and explicitly throw in case the result was `false`. This is also a relatively common pattern in the BCl as well 👍

Does that make sense?

Reply:
It does! Thanks for the walk through, that makes a ton of sense